src/python/pants/backend/codegen/protobuf/scala/rules.py

```diff
@@ -89,58 +131,71 @@ async def generate_scala_from_protobuf(
         Get(
             StrippedSourceFiles, SourceFilesRequest([request.protocol_target[ProtobufSourceField]])
         ),
-        Get(
-            Digest,
-            CreateDigest(
-                [
-                    FileContent(
-                        "protoc-gen-scala",
-                        textwrap.dedent(
-                            f"""\
-                                       #!{bash.path}
-                                       exec {' '.join(jdk_setup.args(bash, tool_classpath.classpath_entries(toolcp_relpath))[1:])} scalapb.ScalaPbCodeGenerator "$@"
-                                       """
-                        ).encode(),
-                        is_executable=True,
-                    )
-                ]
-            ),
-        ),
     )
 
-    unmerged_digests = [
-        all_sources_stripped.snapshot.digest,
-        downloaded_protoc_binary.digest,
-        protoc_gen_scala_digest,
-        empty_output_dir,
-    ]
+    merged_jvm_plugins_digest = EMPTY_DIGEST
+    maybe_jvm_plugins_setup_args = []
+    maybe_jvm_plugins_other_args = []
+    jvm_plugin_artifacts = scalapb.jvm_plugin_artifacts
+    if jvm_plugin_artifacts:
+        materialized_jvm_plugins = await MultiGet(
+            Get(MaterializedJvmPlugin, MaterializeJvmPluginRequest(plugin))
+            for plugin in jvm_plugin_artifacts
+        )
+        jvm_plugin_digests = await MultiGet(
+            Get(Digest, AddPrefix(p.classpath.digest, p.name)) for p in materialized_jvm_plugins
+        )
+        merged_jvm_plugins_digest = await Get(Digest, MergeDigests(jvm_plugin_digests))
+
+        def make_jvm_plugin_arg(plugin):
+            classpath_arg = ":".join(
+                plugin.classpath.classpath_entries(f"{plugins_relpath}/{plugin.name}")
+            )
+            return f"--jvm-plugin={plugin.name}={classpath_arg}"
+
+        for plugin in materialized_jvm_plugins:
+            maybe_jvm_plugins_setup_args.append(make_jvm_plugin_arg(plugin))
+            maybe_jvm_plugins_other_args.append(f"--{plugin.name}_out={output_dir}")
```

Comment:
Putting more of this in the `MaterializedJvmPlugin` class or in the constructing `@rule` would be good, to better isolate the plugin case.

Reply:
Done.